spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/core/PubSubTemplate.java

```diff
@@ -181,19 +195,62 @@ public Subscriber subscribe(String subscription, MessageReceiver messageHandler)
 	}
 
 	@Override
-	public List<PubsubMessage> pull(String subscription, Integer maxMessages, Boolean returnImmediately,
-			RetrySettings retrySettings) {
-		return pull(this.subscriberFactory.createPullRequest(subscription, maxMessages, returnImmediately),
-				retrySettings);
+	public List<PubsubMessage> pull(String subscription, Integer maxMessages,
+			Boolean returnImmediately, RetrySettings retrySettings, boolean acknowledge) {
+		return pull(this.subscriberFactory.createPullRequest(
+				subscription, maxMessages, returnImmediately),
+				retrySettings, acknowledge);
 	}
 
 	@Override
 	public PubsubMessage pullNext(String subscription) {
-		List<PubsubMessage> receivedMessageList = pull(subscription, 1, true, null);
+		List<PubsubMessage> receivedMessageList = pull(subscription, 1, true, null, true);
 
 		return receivedMessageList.size() > 0 ?	receivedMessageList.get(0) : null;
 	}
 
+	public void ackMessage(PubsubMessage message) {
+		ackMessages(Collections.singleton(message), null);
+	}
+
+	public void ackMessages(Collection<PubsubMessage> messages, RetrySettings retrySettings) {
+		SubscriberStub subscriber = this.subscriberFactory.createSubscriberStub(retrySettings);
```

Comment:
We could have a `SubscriberStub` private field if we don't care about the `RetrySettings`, which we currently don't.
I don't know if creating `SubscriberStubs` is expensive. The underlying code is generated and pretty much unreadable, but I'm assuming there must be at least one TCP connection opened to the server, in which case it could make sense to have a subscriber always available?

Reply:
Can we move `RetrySettings` to the constructor, or to just an optional property?
I know it's somewhat limiting to have a single retry settings for all operations, but I think it's a decent compromise.

Reply:
At least that is going to be aligned with Spring `RetryTemplate` configuration we have in other places.

Reply:
Wouldn't it be better to add some `AcknowledgmentCallback` header instead of this suspicious public API?

Yes, that one on its `acknowledge()` may call such a `private` method here in the template, but we wouldn't expose some hook to let end-user to break his/her code.

As a sample I can show `KinesisMessageDrivenChannelAdapter`: https://github.com/spring-projects/spring-integration-aws/blob/master/src/main/java/org/springframework/integration/aws/inbound/kinesis/KinesisMessageDrivenChannelAdapter.java#L1038

We have similar mechanism in Spring Kafka as well: https://github.com/spring-projects/spring-kafka/blob/master/spring-kafka/src/main/java/org/springframework/kafka/support/converter/MessagingMessageConverter.java#L124

However if the story is about `ReceivedMessage` then we don't have choice, but in that case this `ackMessages()` would be much opaque, unless I miss something in those big picture...

Reply:
There are two issues with that:
1) `PubSubMessage` only supports `String` attributes, can't add an `Object` there.
2) We would have to create a new `PubSubMessage` for each new incoming message, in order to add the header to it, which is also what we're trying to avoid here for efficiency.

On this last point, it seems that adding the subscription name to the headers is inevitable, since it is required to acknowledge messages and isn't anywhere in `ReceivedMessage`, so therefore, it also appears that "mutating" the received `PubSubMessage` (e.g., by creating a new one) is really inevitable...

Reply:
`AcknowledgingPubSubMessage` then?
As a wrapper with delegation to the original `PubSubMessage` and extra `Acknowladge` property?

For some other use-case in Spring Integration we have an `MessageWithThreadState`:
```
	private static final class MessageWithThreadState<S> implements Message<Object>, MessageDecorator {

		private final Message<Object> message;

		private final S state;

		@SuppressWarnings("unchecked")
		MessageWithThreadState(Message<?> message, S state) {
			this.message = (Message<Object>) message;
			this.state = state;
		}

		@Override
		public Object getPayload() {
			return this.message.getPayload();
		}

		@Override
		public MessageHeaders getHeaders() {
			return this.message.getHeaders();
		}
...
```

For example

Reply:
But then what would _acknowledge_ the message?
Are you suggesting each message would contain an instance of the Pub/Sub client that can acknowledge? Wouldn't that be really heavy?
If something else acknowledges the message, then that's the current state.

Reply:
One more time: in other projects we provide an `Acknowledgment` abstraction which is populated to the message headers (or some other easy-accessable propety). The implementation of that contract indeed has a reference to some low-level API for the real ack/nack. But that's just a *reference* to the same shared object. How is that heavy?

The current your impl leak the problem that I can have several `PubSubTemplate`s and I won't know which one to use for ack'ing. Meanwhile a clean API with the `Acknowledgment` won't bring me any doubts.

Looks like I again don't have enough arguments to convince you, but thinking one more time about what we have in Spring AMQP, Spring Kafka and Spring Integration (even Spring STOMP) about acknowledging, I see that we have chosen right solution. Everything else is up to you...